src/main/java/org/jenkinsci/plugins/gitclient/verifier/AcceptFirstConnectionVerifier.java

```diff
@@ -27,95 +23,32 @@ public AbstractCliGitHostKeyVerifier forCliGit(TaskListener listener) {
 
     @Override
     public AbstractJGitHostKeyVerifier forJGit(TaskListener listener) {
-        KnownHosts knownHosts;
-        try {
-            knownHosts =
-                    Files.exists(getKnownHostsFile().toPath()) ? new KnownHosts(getKnownHostsFile()) : new KnownHosts();
-        } catch (IOException e) {
-            LOGGER.log(Level.WARNING, e, () -> "Could not load known hosts.");
-            knownHosts = new KnownHosts();
-        }
-        return new AcceptFirstConnectionJGitHostKeyVerifier(listener, knownHosts);
+        return new AcceptFirstConnectionJGitHostKeyVerifier(listener);
     }
 
     public class AcceptFirstConnectionJGitHostKeyVerifier extends AbstractJGitHostKeyVerifier {
 
-        private final TaskListener listener;
-
-        public AcceptFirstConnectionJGitHostKeyVerifier(TaskListener listener, KnownHosts knownHosts) {
-            super(knownHosts);
-            this.listener = listener;
-        }
-
-        @Override
-        public String[] getServerHostKeyAlgorithms(Connection connection) throws IOException {
-            return getPreferredServerHostkeyAlgorithmOrder(connection);
+        public AcceptFirstConnectionJGitHostKeyVerifier(TaskListener listener) {
+            super(listener);
         }
 
         @Override
-        public boolean verifyServerHostKey(
-                String hostname, int port, String serverHostKeyAlgorithm, byte[] serverHostKey) throws Exception {
-            listener.getLogger()
-                    .printf(
-                            "Verifying host key for %s using %s %n",
-                            hostname, getKnownHostsFile().toPath());
-            File knownHostsFile = getKnownHostsFile();
-            Path path = Paths.get(knownHostsFile.getAbsolutePath());
-            String hostnamePort = hostname + ":" + port;
-            boolean isValid = false;
-            if (Files.notExists(path)) {
-                Files.createDirectories(knownHostsFile.getParentFile().toPath());
-                Files.createFile(path);
-                listener.getLogger().println("Creating new known hosts file " + path);
-                writeToFile(knownHostsFile, hostnamePort, serverHostKeyAlgorithm, serverHostKey);
-                isValid = true;
-            } else {
-                KnownHosts knownHosts = getKnownHosts();
-                int hostPortResult = knownHosts.verifyHostkey(hostnamePort, serverHostKeyAlgorithm, serverHostKey);
-                if (KnownHosts.HOSTKEY_IS_OK == hostPortResult
-                        || KnownHosts.HOSTKEY_IS_OK
-                                == knownHosts.verifyHostkey(hostname, serverHostKeyAlgorithm, serverHostKey)) {
-                    isValid = true;
-                } else if (KnownHosts.HOSTKEY_IS_NEW == hostPortResult) {
-                    writeToFile(knownHostsFile, hostnamePort, serverHostKeyAlgorithm, serverHostKey);
-                    isValid = true;
```

Comment:
why not just use [DefaultKnownHostsServerKeyVerifier](https://javadoc.io/static/org.apache.sshd/sshd-core/2.12.1/org/apache/sshd/client/keyverifier/DefaultKnownHostsServerKeyVerifier.html#%3Cinit%3E(org.apache.sshd.client.keyverifier.ServerKeyVerifier,boolean))([ServerKeyVerifier](https://javadoc.io/static/org.apache.sshd/sshd-core/2.12.1/org/apache/sshd/client/keyverifier/ServerKeyVerifier.html) delegate, boolean strict) strict set to false?

Reply:
neither https://javadoc.io/doc/org.eclipse.jgit/org.eclipse.jgit.ssh.apache/latest/org.eclipse.jgit.ssh.apache/org/eclipse/jgit/transport/sshd/SshdSessionFactory.html or https://javadoc.io/doc/org.eclipse.jgit/org.eclipse.jgit.ssh.apache/latest/org.eclipse.jgit.ssh.apache/org/eclipse/jgit/transport/sshd/SshdSessionFactoryBuilder.html offer the possibility to configure this.

I don't like much the idea of having to maintain a re-implementation of this code https://github.com/eclipse-jgit/jgit/blob/5c1c006307cac302b07d42ed3172a3cb3e630dcc/org.eclipse.jgit.ssh.apache/src/org/eclipse/jgit/transport/sshd/SshdSessionFactory.java#L191

Reply:
what about `JGitSshClient.setServerKeyVerifier()` ?   do we need to be configuring the `SshdSessionFactory` directly rather than letting jgitClient do what we want it do with the twaeaks for hosts/credentials?

Reply:
> what about `JGitSshClient.setServerKeyVerifier()` ? do we need to be configuring the `SshdSessionFactory` directly rather than letting jgitClient do what we want it do with the twaeaks for hosts/credentials?

no access to this (it's all static and private) and no configurable (except as said previously having our own implementation `SshdSessionFactory`) FYI this `SshdSessionFactory` creating  `JGitSshClient` (I'm not quite sure what you mean with `jgitClient`)
Again everything is using some statics method without much configuration option such: 
- SshSessionFactory.setInstance
- HttpTransport.setConnectionFactory

look at the branch locally and look at what are the possible methods. configuration are really limited and based on some static methods and system property.

Reply:
```java
class MySshdSessionFactory extends SshdSessionFactory {

  @override
  public SshdSession getSession(URIish uri,
			CredentialsProvider credentialsProvider, FS fs, int tms)
			throws TransportException {
      SshClient client = super.getSession(uri, credentialsProvider, ds, tms);
      client.setServerKeyVerifier(....)    
      ...
   }
}
```

Reply:
```java
	@Override
	public SshdSession getSession(URIish uri,
			CredentialsProvider credentialsProvider, FS fs, int tms)
			throws TransportException {
		SshdSession session = null;
		try {
			session = new SshdSession(uri, () -> {
				File home = getHomeDirectory();
				if (home == null) {
					// Always use the detected filesystem for the user home!
					// It makes no sense to have different "user home"
					// directories depending on what file system a repository
					// is.
					home = FS.DETECTED.userHome();
				}
				File sshDir = getSshDirectory();
				if (sshDir == null) {
					sshDir = new File(home, SshConstants.SSH_DIR);
				}
				HostConfigEntryResolver configFile = getHostConfigEntryResolver(
						home, sshDir);
				KeyIdentityProvider defaultKeysProvider = toKeyIdentityProvider(
						getDefaultKeys(sshDir));
				Supplier<KeyPasswordProvider> keyPasswordProvider = () -> createKeyPasswordProvider(
						credentialsProvider);
				SshClient client = ClientBuilder.builder()
						.factory(JGitSshClient::new)
						.filePasswordProvider(createFilePasswordProvider(
								keyPasswordProvider))
						.hostConfigEntryResolver(configFile) // XXX customize me
						.serverKeyVerifier(new JGitServerKeyVerifier(
								getServerKeyDatabase(home, sshDir))) // XXX customize me
						.signatureFactories(getSignatureFactories())
						.compressionFactories(
								new ArrayList<>(BuiltinCompressions.VALUES))
						.build();
				client.setUserInteraction(
						new JGitUserInteraction(credentialsProvider));
				client.setUserAuthFactories(getUserAuthFactories());
				client.setKeyIdentityProvider(defaultKeysProvider);
				ConnectorFactory connectors = getConnectorFactory();
				if (connectors != null) {
					client.setAgentFactory(
							new JGitSshAgentFactory(connectors, home));
				}
				// JGit-specific things:
				JGitSshClient jgitClient = (JGitSshClient) client;
				jgitClient.setKeyCache(getKeyCache());
				jgitClient.setCredentialsProvider(credentialsProvider);
				jgitClient.setProxyDatabase(proxies);
				jgitClient.setKeyPasswordProviderFactory(keyPasswordProvider);
				String defaultAuths = getDefaultPreferredAuthentications();
				if (defaultAuths != null) {
					jgitClient.setAttribute(
							JGitSshClient.PREFERRED_AUTHENTICATIONS,
							defaultAuths);
				}
				if (home != null) {
					try {
						jgitClient.setAttribute(JGitSshClient.HOME_DIRECTORY,
								home.getAbsoluteFile().toPath());
					} catch (SecurityException | InvalidPathException e) {
					// Ignore
					}
				}
				// Other things?
				return client;
			});
			session.addCloseListener(s -> unregister(s));
			register(session);
			session.connect(Duration.ofMillis(tms));
			return session;
		} catch (Exception e) {
			unregister(session);
			if (e instanceof TransportException) {
				throw (TransportException) e;
			}
			throw new TransportException(uri, e.getMessage(), e);
		}
	}
```

Reply:
yup I wanted to avoid this copy of so much internal code. 
but at least it's still less code duplication than an unmaintained trilead fork :)

Reply:
If this is a desirable way to customize we could send a PR to Apache mina to make the "create the jgitsshclient" a method of the factory rather so it can more easily be overridden to remove the copy paste and leave only the things you want to override.

Reply:
we might need first PR to `org.eclipse.jgit.transport.sshd.SshdSessionFactoryBuilder`.

Reply:
well this will be really crappy.... 
as the method `getSession` to override is using a lot of private methods (to duplicate) 
SshdSession (the one we need to create an instance) is package-private, so it cannot be used. except using the same package.
So turns into:
```
java.lang.SecurityException: class "org.eclipse.jgit.transport.sshd.JenkinsSshdSessionFactory"'s signer ```
information does not match signer information of other classes in the same package
```

tbh that's very ugly..... (probably worst than what is done now)